### PR TITLE
Add ledger support

### PR DESCRIPTION
**Pull request by romaniy173**
Issue: add changes need to implement ledger logic in solution

### CODE REVIEW

File: WalletsLib/CoreHDGroup.h [hunk 1]
```diff
@@ -65,6 +65,7 @@ namespace bs {
 
             virtual void shutdown(void);
             virtual std::set<AddressEntryType> getAddressTypeSet(void) const;
+            std::set<AddressEntryType> getAddressTypeHWSet(void) const;
```
Comment (sergey-chernikov) [inline WalletsLib/CoreHDGroup.h:68]: I think this should be the other way around: you descend a group for HW wallets and override getAddressTypeSet there. Don't duplicate each method for HW wallet.

File: WalletsLib/HDPath.h [hunk 2]
```diff
@@ -55,10 +58,15 @@ namespace bs {
...
 
+         std::vector<Elem>::const_iterator begin() const;
+         std::vector<Elem>::const_iterator end() const;
+
          static Path fromString(const std::string &);
          static Elem keyToElem(const std::string &key);
          static std::string elemToKey(Elem);
 
+
+
```
Comment (sergey-chernikov) [inline WalletsLib/HDPath.h:69]: These 2 empty lines are not needed here, IMO.

File: WalletsLib/CoreHDGroup.h (hunk 1 above)
Comment (romaniy173) [reply WalletsLib/CoreHDGroup.h]: Fixed, thank you

File: WalletsLib/HDPath.h (hunk 2 above)
Comment (romaniy173) [reply WalletsLib/HDPath.h]: Fixed, thank you

File: WalletsLib/CoreHDWallet.cpp [hunk 3]
```diff
@@ -313,7 +313,11 @@ void bs::core::hd::Wallet::createHwStructure(const bs::core::wallet::HwWalletInf
       { AddressEntryType_P2PKH, walletInfo.xpubLegacy_ }
    };
 
-   for (const auto &aet : groupXBT->getAddressTypeHWSet()) {
+   // Adding legacy address support explicitly
+   auto addressTypeSet = groupXBT->getAddressTypeSet();
+   addressTypeSet.insert(AddressEntryType_P2PKH);
```
Comment (sergey-chernikov) [inline WalletsLib/CoreHDWallet.cpp:318]: It's done wrong again. You need an HWGroup class instance where you override getAddressTypeSet method to contain the desirable address types for HWGroup only.

Comment (romaniy173) [reply WalletsLib/CoreHDWallet.cpp]: Thank you, now I got the idea. Done

File: WalletsLib/CoreHDWallet.cpp [hunk 4]
```diff
@@ -222,27 +222,34 @@ std::shared_ptr<hd::Group> hd::Wallet::createGroup(bs::hd::CoinType ct)
...
+         break;
+
+      default:
+         result = std::make_shared<Group>(
+            walletPtr_, ct, netType_, extOnlyFlag_, logger_);
+         break;
+      }
+   }
+   else {
+      result = std::make_shared<HWGroup>(
```
Comment (sergey-chernikov) [inline WalletsLib/CoreHDWallet.cpp:249]: I guess you should forbid the creation of groups for unsupported coin types in HW wallets (e.g. Auth and CC).

Comment (sergey-chernikov) [reply WalletsLib/CoreHDWallet.cpp:249]: Ideally, you should subclass bs::core::hd::Wallet for HW wallet to eliminate all these `if`s and duplication of methods.

Comment (romaniy173) [reply WalletsLib/CoreHDWallet.cpp:253]: Agree with both. I have added exception like a fix for a first comment. The second one is more bigger, we also do not have correct structure for cases like readonly wallet etc, so everything is managing on just flag for now. So I would say I would skip second one and probably will create a ticket to further refactoring of this structure, as far as it not pretty straight forward?

**Pull request merged**